cloudevents/working-drafts/xml-format.md

```diff
@@ -0,0 +1,290 @@
+# XML Event Format for CloudEvents
+
+## Abstract
+
+This format specification for CloudEvents defines how events are expressed
+in XML documents.
+
+## Status of this document
+
+This document is a working draft.
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+2. [Attributes](#2-attributes)
+3. [Data](#3-data)
+4. [Envelope](#4-envelope)
+5. [XML Batch Format](#5-xml-batch-format)
+6. [Examples](#6-examples)
+
+## 1. Introduction
+
+[CloudEvents][ce-spec] is a standardized and protocol-agnostic definition of the
+structure and metadata description of events. This specification defines how
+elements defined in the CloudEvents specification are to be represented using
+[Extensible Markup Language (XML)](xml-spec) documents.
+
+* The [Attributes](#2-attributes) section describes the representation and
+data type mappings for CloudEvents context attributes.
+
+* The [Data](#3-data) section defines the container for the data portion of a
+CloudEvent.
+
+* The [Envelope](#4-envelope) section defines an XML element for CloudEvents
+attributes and an associated media type.
+
+* The [Batch](#5-xml-batch-format) section describes how multiple CloudEvents
+can be packaged into a single XML document.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Approach
+
+This XML representation used is deliberately verbose in favor of readability
+and deterministic processing.
+
+Preservation of type information for CloudEvent attributes is supported allowing
+custom extensions to be communicated without type loss.
+
+A schema-less approach has been taken favoring convention over rigid document
+structure.
+
+The namespace `http://cloudevents.io/xmlformat/V1` MUST be used.
+
+When namespace prefixes are used a prefix of `ce` is preferred but MUST NOT
+be expected from an XML document processing perspective.
+
+An XML document preamble SHOULD be included to ensure deterministic processing.
+
+## 2. Attributes
+
+The CloudEvents type system is mapped to the XML schema types as follows :
+
+| CloudEvents   |  XML Schema Type | Notes |
+| :-------------| :--------------- | :---- |
+| Boolean       | [xs:boolean][xml-primitives] | |
+| Integer       | [xs:int][xml-primitives] | |
+| String        | [xs:string][xml-primitives] | |
+| Binary        | [xs:base64Binary][xml-primitives] | |
+| URI           | [xs:anyURI][xml-primitives] following [RFC 3986][rfc3986]| |
+| URI-reference | [xs:anyURI][xml-primitives] following [RFC 3986][rfc3986] | |
```

Comment:
Ooh, that's a shame... this means we can't *really* preserve type information for extension attributes, as we can't tell the difference between URI and URI-reference.

Perhaps we should have an extra attribute for URI-reference of "ref=true" or something like that?

Reply:
Yep .. I struggle with this a bit as well ...

The only really way out of this is to define our own set of 'ce' types and use those - that's the road I was going down when I was playing with a schema.

So you'd end up with something like
<myextension xsi:type="ce:string">i'm a string</myextension>

It's really no different that using the pre-defined XS types - I'm happy to make this switch, I don't think it's a major change

Reply:
There's one benefit to using predefined xs types, which is that some editors support them - for example, for my unit tests I've got some written in code rather than as XML, because when I try to put an invalid value in an `xs:boolean` element, Visual Studio emits a warning. But I agree that in other ways it's cleaner than having an extra attribute.

So some options:

- Go with the spec as written, unable to differentiate between URI and URI-Reference
- Introduce `ce:uri` but leave the rest of the `xs:` types as they are
- Stick with just `xs:` types but add a qualifying attribute to go alongside `xs:anyURI`
- Ditch the `xs` types entirely and go full-on `ce:` types. (I don't know whether using `xsi:type` for that becomes a problem)

Smarter folks than me may well have other suggestions of course.

Reply:
Spec has been updated to reflect `ce` types for the context attributes and `xs` for the data content.